lib/netlink.c

```diff
@@ -0,0 +1,465 @@
+	{
+		/* There can be race, we do not return error here */
+		/* FIXME I18N: *"unknown"* is too generic. Use context. */
+		/* TRANSLATORS: unknown network interface, maximum 15
+		 * (IF_NAMESIZE-1) bytes */
+		nl->addr.ifname = _("unknown");
+	}
+	nl->addr.ifa_flags = (uint32_t)(ifaddr->ifa_flags);
+	/* If IFA_CACHEINFO is not present, suppose permanent addresses. */
+	nl->addr.ifa_valid = UINT32_MAX;
+	ON_DBG(ADDR, dbg_addr(nl));
+
+	/* Process rtattr. */
+	len = nh->nlmsg_len - NLMSG_LENGTH(sizeof(*ifaddr));
+	for (attr = IFA_RTA(ifaddr); RTA_OK(attr, len);
+	     attr = RTA_NEXT(attr, len)) {
+		/* Proces most common rta attributes */
+		DBG(ADDR, ul_debugobj(attr, "processing rtattr"));
+		switch (attr->rta_type) {
+		case IFA_ADDRESS:
+			nl->addr.ifa_address = RTA_DATA(attr);
+			nl->addr.ifa_address_len = RTA_PAYLOAD(attr);
+			if (!has_local_address) {
+				nl->addr.address = RTA_DATA(attr);
+				nl->addr.address_len = RTA_PAYLOAD(attr);
+			}
+			DBG(ADDR,
+			    ul_debug(" IFA_ADDRESS%s: %s",
+				     (has_local_address ? "" :
+				      " (setting address)"),
+				     ul_nl_addr_ntop_ifa_address(&(nl->addr))));
+		break;
+		case IFA_LOCAL:
+			/* Point to Point interface listens has local address
+			 * and listens there */
+			has_local_address = true;
+			nl->addr.ifa_local = nl->addr.address = RTA_DATA(attr);
+			nl->addr.ifa_local_len =
+				nl->addr.address_len = RTA_PAYLOAD(attr);
+			DBG(ADDR,
+			    ul_debug(" IFA_LOCAL (setting address): %s",
+				     ul_nl_addr_ntop_ifa_local(&(nl->addr))));
+		break;
+		case IFA_CACHEINFO:
+		{
+			struct ifa_cacheinfo *ci =
+				(struct ifa_cacheinfo *)RTA_DATA(attr);
+			nl->addr.ifa_valid = ci->ifa_valid;
+			DBG(ADDR,
+			    ul_debug(" IFA_CACHEINFO: ifa_prefered = %u,"
+				     "ifa_valid = %u",
+				     nl->addr.ifa_prefered,
+				     nl->addr.ifa_valid));
+		}
+		break;
+		case IFA_FLAGS:
+			nl->addr.ifa_flags = *(uint32_t *)(RTA_DATA(attr));
+			DBG(ADDR, ul_debug(" IFA_FLAGS: 0x%08x",
+					  nl->addr.ifa_flags));
+		break;
+		default:
+			DBG(ADDR, ul_debug(" rta_type = 0x%04x",
+					  attr->rta_type));
+		break;
+		}
+	}
+	DBG(NLMSG, ul_debugobj(nl, "callback %p", nl->callback_addr));
+	return (*(nl->callback_addr))(nl);
+}
+
+static int process_msg(struct ul_nl_data *nl, struct nlmsghdr *nh)
+{
+	int rc = 0;
+
+	nl->rtm_event = UL_NL_RTM_DEL;
+	switch (nh->nlmsg_type) {
+	case RTM_NEWADDR:
+		nl->rtm_event = UL_NL_RTM_NEW;
+		/* fallthrough */
+	case RTM_DELADDR:
+	/* If callback_addr is not set, skip process_addr */
+		DBG(NLMSG, ul_debugobj(nl, "%s",
+				       (UL_NL_IS_RTM_DEL(nl) ?
+					"RTM_DELADDR" : "RTM_NEWADDR")));
+		if (nl->callback_addr)
+			rc = process_addr(nl, nh);
+		break;
+	/* More can be implemented in future (RTM_NEWLINK, RTM_DELLINK etc.). */
+	default:
+		DBG(NLMSG, ul_debugobj(nl, "nlmsg_type = %hu", nh->nlmsg_type));
+		break;
+
+	}
+	return rc;
+}
+
+int ul_nl_process(struct ul_nl_data *nl, bool async, bool loop)
+{
+	char buf[4096];
```

Comment:
4096 -> BUFSIZ ?

Reply:
Actually, there is no exact size defined. If we don't want to handle MSG_TRUNC, it should cover sizeof (struct nlmsghdr) plus the expected message payload size. BUFSIZ (8192) is surely more safe than 4096.